engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift

```diff
@@ -0,0 +1,116 @@
+  case important
+
+  /// Messages indicating a potential issue or an unexpected situation that isn't critical.
+  case warning
+
+  /// Messages indicating a runtime error from which the application can potentially recover.
+  case error
+
+  /// Messages indicating a critical condition. Causes the application to immediately terminate.
+  case fatal
+}
+
+/// A singleton logger for outputting runtime messages.
+///
+/// This logger allows for messages to be logged at different severity levels. Its output can be
+/// filtered by setting the `logLevel` property to the minimum log level to be logged.
+///
+/// **Usage:**
+/// ```swift
+/// Logger.logInfo("Application has started.")
+/// Logger.instance.logLevel = .warning // Only show warnings and above
+/// Logger.logError("Failed to load asset: \(assetKey)")
+/// ```
+@objc(FlutterLogger) public final class Logger: NSObject {
+  /// The shared singleton instance of the logger.
+  @objc public static let instance = Logger()
+
+  /// Logs a message at `LogLevel.info`.
+  @objc public static func logInfo(_ message: String) {
+    instance.log(level: .info, message)
+  }
+
+  /// Logs a message at `LogLevel.important`.
+  @objc public static func logImportant(_ message: String) {
+    instance.log(level: .important, message)
+  }
+
+  /// Logs a message at `LogLevel.warning`.
+  @objc public static func logWarning(_ message: String) {
+    instance.log(level: .warning, message)
+  }
+
+  /// Logs a message at `LogLevel.error`.
+  @objc public static func logError(_ message: String) {
+    instance.log(level: .error, message)
+  }
+
+  /// Logs a message at `LogLevel.fatal` and immediately terminates the application.
+  @objc public static func logFatal(_ message: String) {
+    instance.log(level: .fatal, message)
+    abort()
+  }
+
+  /// Logs a message unconditionally.
+  @objc public static func logDirect(_ message: String) {
+    instance.outputWriter.writeLine(message)
+  }
+
+  /// Sets the minimum log level.
+  @objc public static func setLogLevel(_ level: LogLevel) {
+    instance.logLevel = level
+  }
+
+  var logLevel = LogLevel.info
+  let outputWriter: OutputWriter
+
+  private override init() {
+#if os(iOS)
+    outputWriter = SyslogOutputWriter()
+#elseif os(macOS)
+    outputWriter = StdoutOutputWriter()
+#endif
+  }
+
+  private func log(level: LogLevel, _ message: String) {
+    if level.rawValue >= logLevel.rawValue {
+      outputWriter.writeLine(message)
+    }
+  }
+}
+
+protocol OutputWriter {
+  func writeLine(_ message: String)
+}
+
+final class SyslogOutputWriter: OutputWriter {
+  func writeLine(_ message: String) {
+    // TODO(cbracken): replace this with os_log-based approach.
+    // https://github.com/flutter/flutter/issues/44030
+    message.withCString { vsyslog(LOG_ALERT, "%s", getVaList([$0])) }
+  }
+}
+
+final class StdoutOutputWriter: OutputWriter {
+  func writeLine(_ message: String) {
+    fputs(message, stdout)
+    fputs("\n", stdout)
+    fflush(stdout)
```

Comment:
can we just do `print`? https://developer.apple.com/documentation/swift/print(_:separator:terminator:)

Reply:
Print has platform-specific behaviour and is not documented to write directly to `stdout`. You can think of this as implementing a peer to `print`.